core/math/color.h

```diff
@@ -196,15 +196,15 @@ struct Color {
 	_FORCE_INLINE_ bool operator<(const Color &p_color) const; //used in set keys
 	operator String() const;
 
-	//for binder
+	// For the binder.
 	_FORCE_INLINE_ void set_r8(int32_t r8) { r = (CLAMP(r8, 0, 255) / 255.0); }
-	_FORCE_INLINE_ int32_t get_r8() const { return CLAMP(uint32_t(r * 255.0), 0, 255); }
+	_FORCE_INLINE_ int32_t get_r8() const { return MIN(uint32_t(r * 255.0), 255u); }
```

Comment:
Using `255u` to solve this:
```
 ./core/math/color.h:207:49:{207:49-207:78}{207:49-207:78}: error: comparison of integers of different signs: 'typeof (uint32_t(this->a * 255.))' (aka 'unsigned int') and 'typeof (255)' (aka 'int') [-Werror,-Wsign-compare,2]
        _FORCE_INLINE_ int32_t get_a8() const { return MIN(uint32_t(a * 255.0), 255); }
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode_12.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS14.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:572:22:{572:22-572:53}{572:22-572:53}: note: expanded from macro 'MIN'
     #define MIN(A,B) __NSMIN_IMPL__(A,B,__COUNTER__)
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode_12.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS14.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:571:144:{571:123-571:143}{571:146-571:166}: note: expanded from macro '__NSMIN_IMPL__'
     #define __NSMIN_IMPL__(A,B,L) ({ __typeof__(A) __NSX_PASTE__(__a,L) = (A); __typeof__(B) __NSX_PASTE__(__b,L) = (B); (__NSX_PASTE__(__a,L) < __NSX_PASTE__(__b,L)) ? __NSX_PASTE__(__a,L) : __NSX_PASTE__(__b,L); })
                                                                                                                           ~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~
```

We could also use `CLAMP(int32_t(r * 255.0), 0, 255)` if we prefer to use `int32_t` directly instead of `uint32_t`. But I guess that would truncate potential negative `r`s (which shouldn't happen in theory) instead of wrap around.

Reply:
Negative values for colors don't actually exist because negative light doesn't exist. However, there are use cases where it makes sense in code, for example adding a color of `(-0.1, 0.2, 0.3)` would change the color by removing 10% red, adding 20% green, and adding 30% blue.

Not sure what's the desired behavior, but I guess there isn't much harm in supporting this. Currently the C# code does support this, since `r8` etc are signed ints with no clamping, so it's currently inconsistent with the C++ code.

Reply:
It's a bit abusing the `Color` data type to present a `ΔColor` (i.e. `Color2 - Color1`) though, which has different mathematical properties, just like `K` (Kelvin) and `ΔK` have different properties.

I don't think that making the `Color` boundaries flexible to support deltas is a good idea, as it means you can actually end up with negative Colors in rendering code, which breaks a lot of assumptions.

If you want to do the above, the right way would be `- (0.1, 0, 0) + (0, 0.2, 0.3)`. Might be less elegant but it doesn't open up a can worms :)

Reply:
> If you want to do the above, the right way would be `- (0.1, 0, 0) + (0, 0.2, 0.3)`

That *technically* still uses a negative color though, even though it's not saved to a variable.

Anyway... for `r8`/etc specifically, I just realized, if we want to enforce 0 to 255, why not simply cast to `uint8_t`?

Reply:
Actually `Color` doesn't do clamping in its operators or constructor, so it's valid to have negative colors (or colors with > 255, which is the case for HDR). These clamped setters and getters seem specifically about the binding code, I'm not sure what requires the additional restriction and use of `int32_t`, but I guess @reduz had a good reason for not going with something simpler.

Reply:
IRC discussion around this: https://freenode.logbot.info/godotengine-devel/20201109#c5761543-c5761781